lib/ukdebug/gdbstub.c

```diff
@@ -1106,3 +1113,65 @@ int gdb_dbg_trap(int errnr, struct __regs *regs)
 
 	return r;
 }
+
+static int gdb_setup_backing_console(void)
+{
+	/* Ignore breakpoints if the stub is not enabled on the command line. */
+	if (gdb_backing_console_id == GDB_BACKING_CONSOLE_NONE)
+		return -ENXIO;
+
+	/* Ignore breakpoints if the backing console is not registered.
+	 * Maybe the backing console will be initialized later.
+	 */
+	if (gdb_backing_console_id >= uk_console_count()) {
+		uk_pr_warn("Console device %" __PRIu16 " not available\n",
+			   gdb_backing_console_id);
+		return -ENXIO;
+	}
+
+	gdb_backing_console = uk_console_get(gdb_backing_console_id);
+	UK_ASSERT(gdb_backing_console);
+
+	uk_console_init(&gdb_virt_console, "GDB virtual console",
+			&gdb_virt_ops, UK_CONSOLE_FLAG_STDOUT);
+	if (gdb_virt_console_on()) {
+		uk_console_register(&gdb_virt_console);
+	} else if (gdb_virt_console_auto()) {
+		/* Must check before diabling the flags (see below) */
+		if (gdb_backing_console->flags)
+			uk_console_register(&gdb_virt_console);
+	}
+
+	/* If we didn't disable the STDOUT and STDIN flags, calls to
+	 * `uk_console_[out|in]` that happen outside of the gdb stub
+	 * could mess up the gdb communications. E.g., a call to
+	 * `uk_console_out` could write some bytes to the device that
+	 * look like garbage to the gdb host on the other end.
+	 * We assume here that the backing device is registered with
+	 * `ukconsole` exactly  once. This means we can also assume that
+	 * disabling the STDOUT and STDIN flags makes sure that the
+	 * underlying device is never used except explicitly stated with
+	 * `uk_console_[out|in]_with(gdb_backing_console, ...)`.
+	 */
+	gdb_backing_console->flags &= ~UK_CONSOLE_FLAG_STDOUT;
+	gdb_backing_console->flags &= ~UK_CONSOLE_FLAG_STDIN;
+
+	return 0;
+}
+
+static int gdb_late_breakpoint(struct uk_init_ctx *ctx __unused)
+{
+	/* This is the first time that the stub is entered. */
+	UK_ASSERT(!gdb_backing_console);
+
+	uk_pr_info("Waiting for debugger ...\n");
```

Comment:
Can we print the console device number as well, like:
`"Waiting for debugger connection at console %"__PRIu16"...\n"`